### PR TITLE
Enable Model Parallel support for torch via DTensor integration

### PR DESCRIPTION
**Pull request by buildwithsuhana**
## Description
This PR implements Model Parallel (MP) support for the PyTorch backend, integrating Keras 3 ModelParallel distributions with PyTorch's native DTensor (Distributed Tensor) infrastructure. This allows for sharding model weights and intermediate activations across multiple devices.

### Key Implementations:

1. DTensor Integration for Model Weights
      - Implemented distribute_variable and distribute_tensor to map Keras TensorLayout and DeviceMesh to Torch's Placement and DeviceMesh objects.
      - Extended Variable initialization to support layout-aware creation, allowing weights to be sharded or replicated across the mesh during the model's build phase.

2. Model Parallel Data Distribution
      - Updated TorchEpochIterator and _distribute_data to support ModelParallel layouts.
      - Data is automatically sharded or replicated across the device mesh according to the active distribution's data layout.

3. DTensor-Aware Operations
      - Implemented _promote_to_dtensor in the Torch backend to handle operations involving mixed local tensors and DTensors.
      - Ensured ops like maximum, minimum, and dot_product_attention properly promote inputs to maintain distribution consistency during the forward pass.

This PR has changes from the Add Data Parallel support for Torch backend PR (https://github.com/keras-team/keras/pull/23316) as well and needs to be rebased once this PR is merged.
 
## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/backend/torch/trainer.py [hunk 1]
```diff
@@ -42,10 +46,11 @@ def train_step(self, data):
         x, y, sample_weight = data_adapter_utils.unpack_x_y_sample_weight(data)
 
         # Compute predictions
+        model = getattr(self, "ddp_model", self)
         if self._call_has_training_arg:
-            y_pred = self(x, training=True)
+            y_pred = model(x, training=True)
         else:
-            y_pred = self(x)
+            y_pred = model(x)
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: What about `test_step` and `predict_step`?

File: keras/src/backend/torch/trainer.py [hunk 2]
```diff
@@ -113,6 +118,40 @@ def make_train_function(self, force=False):
...
+            object.__setattr__(
+                self,
+                "ddp_model",
+                DistributedDataParallel(
+                    self,
+                    device_ids=device_ids,
+                    process_group=process_group,
+                    find_unused_parameters=False,
+                ),
+            )
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: Looking at this code, it looks like it's creating `ddp_model` via `DistributedDataParallel` even in the `ModelParallel` case.

Am I misreading this? Or is this intentional? If so, why?

File: keras/src/backend/torch/trainer.py [hunk 3]
```diff
@@ -144,6 +183,59 @@ def test_function(data):
 
         self.test_function = test_function
 
+    def _sync_metrics(self):
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: I think it would be cleaner to override `get_metrics_result` for this:

```python
def get_metrics_result(self):
    # The code you have in `_sync_metrics` right now here
    # ...
    return super().get_metrics_results()
```

File: keras/src/backend/torch/core.py [hunk 4]
```diff
@@ -119,6 +137,10 @@ def _initialize(self, value):
             ).to(get_device())
 
     def _direct_assign(self, value):
+        if self._layout is not None:
+            from keras.src.backend.torch import distribution_lib
+
+            value = distribution_lib.distribute_tensor(value, self._layout)
         with torch.no_grad():
             self.value.copy_(value)
```
Comment (hertschuh) [inline keras/src/backend/torch/core.py]: Will `self.value.copy_` keep the distribution of `self.value` or the distribution of `value`? Or does it not matter?

**Review by hertschuh — COMMENTED** [review]
Would it be possible to split this into 2 PR?
- DataParallel support first
- ModelParallel next

I'd like to really understand which parts are needed for DDP. And I'm a bit confused by the fact that some code appears to be needed for both when I thought there was basically no overlap between the DataParallel implementation (based on `DistributedDataParallel`) and the ModelParallel implementation based on `DTensor`.

File: keras/src/backend/torch/trainer.py (hunk 3 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py]: Done

File: keras/src/backend/torch/trainer.py (hunk 2 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py]: Updated the implementation to include an explicit check for this: `if active_distribution is None or isinstance(active_distribution, DataParallel):`

File: keras/src/backend/torch/trainer.py (hunk 1 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/trainer.py]: Updated

File: keras/src/backend/torch/core.py (hunk 4 above)
Comment (buildwithsuhana) [reply keras/src/backend/torch/core.py]: It keeps the distribution of self.value. In PyTorch, copy_ is an in-place update that fills the existing variable with new data while preserving its original layout and sharding.

**Comment by buildwithsuhana** [comment]
> Would it be possible to split this into 2 PR?
> 
> * DataParallel support first
> * ModelParallel next
> 
> I'd like to really understand which parts are needed for DDP. And I'm a bit confused by the fact that some code appears to be needed for both when I thought there was basically no overlap between the DataParallel implementation (based on `DistributedDataParallel`) and the ModelParallel implementation based on `DTensor`.

Split into 2 PRs (PR with Data Parallel: https://github.com/keras-team/keras/pull/23316)
While DDP (for parameters/gradients) and DTensor (for variable sharding) are different technologies, data parallelism exists in both. Handling the statistics of that data parallelism (metrics and device coordination) is why the code overlaps a bit.
